metaflow/plugins/aws/awslambda/lambda_decorator.py

```diff
@@ -0,0 +1,264 @@
+        from metaflow.package import MetaflowPackage
+        from metaflow.flowspec import FlowSpec
+        from metaflow.datastore import MetaflowDataStore
+        from metaflow.metaflow_environment import MetaflowEnvironment
+        from metaflow.event_logger import EventLogger
+
+from metaflow.metaflow_config import (
+    LAMBDA_IMAGE_URI,
+    LAMBDA_ROLE_ARN,
+)
+
+DEFAULT_TIMEOUT_SECONDS = 600
+LAMBDA_DEFAULT_MEMORY_MB = 2048
+
+
+class LambdaException(MetaflowException):
+    headline = "AWS Lambda error"
+
+
+class LambdaDecorator(StepDecorator):
+    """
+    The decorator is called @awslambda to avoid confusion, since "lambda" is a reserved
+    word in Python.
+    """
+
+    name = "awslambda"
+    defaults = {
+        "image": None,
+        "refresh_image": False,
+    }
+
+    package_url = None  # type: Optional[str]
+    package_sha = None  # type: Optional[str]
+    memory_mb = LAMBDA_DEFAULT_MEMORY_MB  # type: int
+    run_time_limit = None  # type: Optional[int]
+    ds_root = None  # type: Optional[str]
+    lambda_arn = None  # type: Optional[str]
+    lambda_name = None  # type: Optional[str]
+
+    def step_init(
+        self,
+        flow,  # type: "FlowSpec"
+        graph,  # type : "FlowGraph"
+        step,  # type: str
+        decos,  # type: List[Decorator]
+        environment,  # type: "MetaflowEnvironment"
+        datastore,  # type: "MetaflowDataStore"
+        logger,  # type: "Callable[[str], None]"
+    ):
+        if datastore.TYPE != "s3":
+            raise LambdaException("The *@awslambda* decorator requires --datastore=s3.")
+
+        self.logger = logger
+        self.environment = environment
+        self.step = step
+
+        for deco in decos:
+            if isinstance(deco, CondaStepDecorator):
+                raise LambdaException(
+                    "*@awslambda* doesn't support using @conda. "
+                    "We recommend using custom docker images to manage"
+                    " dependencies instead"
+                )
+
+            if isinstance(deco, BatchDecorator):
+                raise LambdaException(
+                    "You can't use @batch and *@awslambda* at the same time"
+                )
+
+            if isinstance(deco, ResourcesDecorator):
+                for k, v in deco.attributes.items():
+                    if k == "memory":
+                        if float(v) > LAMBDA_MAX_MEMORY_MB:
+                            raise LambdaException(
+                                "{r} cannot exceed {max_v} for *@awslambda* (currently set to {v})".format(
+                                    r=k, max_v=LAMBDA_MAX_MEMORY_MB, v=v
+                                )
+                            )
+                        self.memory_mb = v
+
+        # Run time limit more than 15m is effectively ignored
+        self.run_time_limit = min(
+            get_run_time_limit_for_task(decos, default_limit=DEFAULT_TIMEOUT_SECONDS),
```

Comment:
Curious what happens if we leave the run-time limit to 5 days? Does AWS Lambda complain?

Reply:
Also, when does the run-time-limit clock start?

Reply:
Yeah you can't set it above 900 seconds. Starts when Lambda is invoked.